gluster-block: implement expand feature interface

pkalever opened the pull request:
<!--
Thanks for sending a pull request! Your contribution is appreciated.

Here are some tips for you:

1. Read the contributing guide at [https://github.com/heketi/heketi/blob/master/docs/contributing.md]
2. Split the changes up into minimal and atomic commits.  [https://github.com/heketi/heketi/blob/master/docs/contributing.md#splitting-your-change-into-commits]
3. Write meaningful commit messages [https://github.com/heketi/heketi/blob/master/docs/contributing.md#good-commit-messages]
4. Test your changes: run `make test`.  [https://github.com/heketi/heketi/blob/master/docs/contributing.md#testing-the-change]
5. Write a meaningful PR text. Remember: one PR per feature or bugfix. If in doubt, split your patchset into multiple PRs.
-->

### What does this PR achieve? Why do we need it?
Implement block volume expand interface in heketi

### Notes for the reviewer

Pushing it live with the intention that I will have my code side review comments ready by that time I finish on the TODO list.

The RFE just works fine now but also check for TODO list at the end of this comment which I got from John and Talur in a demo meeting.


```
Before Expand:
-------------
[root@server1 ~]# heketi-cli blockvolume create --size 1
Name: blockvol_365f428f2bf4b5a37f644c239f148058
Size: 1
Volume Id: 365f428f2bf4b5a37f644c239f148058
Cluster Id: 0b5740b0e3c34e935c1cd73e8bbf3c71
Hosts: [192.168.121.42 192.168.121.68 192.168.121.155]
IQN: iqn.2016-12.org.gluster-block:afcc41a3-1c46-4102-812b-5a0767cfb16a
LUN: 0
Hacount: 3
Username:
Password:
Block Hosting Volume: 7b3e7fceae17a57e0fc9ddedcce75912

[root@server1 ~]# heketi-cli volume info 7b3e7fceae17a57e0fc9ddedcce75912 | grep "Free Size"
Free Size: 8

[root@server1 ~]# targetcli ls | grep vol_7b3e7fceae17a57e0fc9ddedcce75912
 o- blockvol_365f428f2bf4b5a37f644c239f148058 [vol_7b3e7fceae17a57e0fc9ddedcce75912@... (1.0GiB) activated]

After Expand:
------------
[root@server1 ~]# heketi-cli blockvolume expand --block-volume 365f428f2bf4b5a37f644c239f148058 --new-size 2
Name: blockvol_365f428f2bf4b5a37f644c239f148058
Size: 2
Volume Id: 365f428f2bf4b5a37f644c239f148058
Cluster Id: 0b5740b0e3c34e935c1cd73e8bbf3c71
Hosts: [192.168.121.42 192.168.121.68 192.168.121.155]
IQN: iqn.2016-12.org.gluster-block:afcc41a3-1c46-4102-812b-5a0767cfb16a
LUN: 0
Hacount: 3
Username:
Password:
Block Hosting Volume: 7b3e7fceae17a57e0fc9ddedcce75912

[root@server1 ~]# heketi-cli volume info 7b3e7fceae17a57e0fc9ddedcce75912 | grep "Free Size"
Free Size: 7

[root@server1 ~]# targetcli ls | grep vol_7b3e7fceae17a57e0fc9ddedcce75912
 o- blockvol_365f428f2bf4b5a37f644c239f148058 [vol_7b3e7fceae17a57e0fc9ddedcce75912@... (2.0GiB) activated]

```

TODO:
1. Implement Clean() and CleanDone() methods  --> Done with V2
2. Add unit tests
3. defer the block volume delete in the functional test case --> Done with V2

pkalever commented:
@phlogistonjohn @raghavendra-talur can we request the CI for running the Smoke tests? or does it already got triggered automatically?

phlogistonjohn commented:
/ok to test

phlogistonjohn commented:
Added 'do not merge' label for now, once TODOs are done we will remove the label.

pkalever commented:
@phlogistonjohn @raghavendra-talur I have added new patches to this series, implementing Rollback, clean and CleanDone methods, also moved BHV delete in functional tests to a defer method. PTAL.

phlogistonjohn commented:
Travis CI run is failing due to a go vet error. This appears to be a real issue, please investigate.

Centoc-CI run failed in TestBlockVolumeExpand. This will need investigation too.
```
--- FAIL: TestBlockVolumeExpand (66.95s)
    assert.go:44:
        ^MASSERT:»»»func (github.com/heketi/heketi/tests/functional/TestSmokeTest/tests.TestBlockVolumeExpand) 0x71f629
        ^M»»File /root/go/src/github.com/heketi/heketi/tests/functional/TestSmokeTest/tests/block_volume_test.go:283
        ^M»»Info: [expected volInfo.Size == 146 got: 200]
```

I will begin reviewing too.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,264 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+		// to the db.
+		// This is only noteworthy because it is different from regular
+		// volumes which determines everything up front. Here certain
+		// values are determined by gluster-block commands.
```
We're probably copying the code for file volume expand a bit too closely here. This isn't the best pattern, at least as far as mutating the entry objects outside of a transaction and then expecting consistent results afterward.
A better pattern would be to capture the results needed from the system in the expand exec function in an object(s) or attributes specifically created to track what the state of the system is rather than piggyback on db entries.
This also makes it cleaner wrt removing the *<x>Entry as an argument to initialize the operation.

pkalever replied:
John,

I have copied this piece of comments intentionally from the same file, it is part of function:
```
func (bvc *BlockVolumeCreateOperation) Finalize() error {
    ...
}
```

I was actually unclear with what we wanted to do to fix completely.

Hence, I would like to let this PR carry as it is, if its not a serious problem, and I will definitely address it in both the places in a separate PR.

pkalever commented on apps/glusterfs/app_block_volume.go:
```diff
@@ -176,3 +176,78 @@ func (a *App) BlockVolumeDelete(w http.ResponseWriter, r *http.Request) {
+			return err
+		}
+
+		if msg.Size == blockvolume.Info.Size {
```
Ok

pkalever commented on apps/glusterfs/block_volume_entry.go:
```diff
@@ -320,6 +320,65 @@ func (v *BlockVolumeEntry) destroyFromHost(
 	return nil
 }
 
+func (v *BlockVolumeEntry) InfoBlockVolumeFromHost(executor executors.Executor,
```
ok

pkalever commented on apps/glusterfs/block_volume_entry.go:
```diff
@@ -320,6 +320,65 @@ func (v *BlockVolumeEntry) destroyFromHost(
+	return e
+}
+
+func (v *BlockVolumeEntry) expandBlockVolumeExec(executor executors.Executor,
```
ok

pkalever replied:
Yes, for now, it only called by Exec(), I will call the executor function directly.

pkalever commented on apps/glusterfs/errors.go:
```diff
@@ -25,6 +25,7 @@ var (
 	ErrKeyExists        = errors.New("Key already exists in the database")
 	ErrNoReplacement    = errors.New("No Replacement was found for resource requested to be removed")
 	ErrCloneBlockVol    = errors.New("Cloning of block hosting volumes is not supported")
+	ErrInvalidRequest   = errors.New("Invalid request")
```
John, I had looked at the existing errors and none of them looks like a match to give a better hint in this case, if you have any thoughts you can share here, so that I will drop it.

This description I had copied from errno.h
```
EINVAL          Invalid argument (POSIX.1-2001).
```

which looked more generic, I can tune it specific to my functionality of code, but the reusability will be compromised.

Let me know what you think.

pkalever commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,264 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+			return err
+		}
+		bve.bvol = v
+		if bve.bvol.Pending.Id != "" {
```
ok

pkalever commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,264 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+// with the given blockvolume entry, db connection and newsize (in GB) that the
+// blockvolume is to be expanded by.
+func NewBlockVolumeExpandOperation(
+	bvol *BlockVolumeEntry, db wdb.DB, newSize int) *BlockVolumeExpandOperation {
```
ok

pkalever commented:
V3:
Addressed:
```
--- FAIL: TestBlockVolumeExpand (66.95s)
    assert.go:44:
        ^MASSERT:»»»func (github.com/heketi/heketi/tests/functional/TestSmokeTest/tests.TestBlockVolumeExpand) 0x71f629
        ^M»»File /root/go/src/github.com/heketi/heketi/tests/functional/TestSmokeTest/tests/block_volume_test.go:283
        ^M»»Info: [expected volInfo.Size == 146 got: 200]
```

The rest of the comments are addressed in a separate patch.

PTAL


Thanks!

pkalever commented:
V3:
Addressed:

- Travis CI run is failing due to a go vet error. 
- Centoc-CI run failed in TestBlockVolumeExpand.
- The rest of the comments are addressed in a separate patch
    - Moved multiple size checks to Build()
    - Change function name from InfoBlockVolumeFromHost() -> BlockVolumeInfoFromHost()
    - Discarded expandBlockVolumeExec() and called the executor directly from Exec()
    - Used IDs as arguments rather than *Entry types at NewBlockVolumeExpandOperation()
    - Added check for pending BHV along with existing pending BV in Build()

PTAL


Thanks!

pkalever commented:
V3:
Addressed:

- Travis CI run is failing due to a go vet errors and later by go fmt errors
- Centoc-CI run failed in TestBlockVolumeExpand.
- The rest of the comments are addressed in a separate patch
    - Moved multiple size checks to Build()
    - Change function name from InfoBlockVolumeFromHost() -> BlockVolumeInfoFromHost()
    - Discarded expandBlockVolumeExec() and called the executor directly from Exec()
    - Used IDs as arguments rather than *Entry types at NewBlockVolumeExpandOperation()
    - Added check for pending BHV along with existing pending BV in Build()

PTAL


Thanks!

pkalever commented:
Rebased on the latest master.

phlogistonjohn commented on executors/cmdexec/block_volume.go:
```diff
@@ -192,3 +194,163 @@ func (c *CmdExecutor) ListBlockVolumes(host string, blockhostingvolume string) (
+	}
+}
+
+func (c *CmdExecutor) InfoBlockVolume(host string, blockhostingvolume string,
```
BlockVolumeInfo?

phlogistonjohn commented on executors/cmdexec/block_volume.go:
```diff
@@ -192,3 +194,163 @@ func (c *CmdExecutor) ListBlockVolumes(host string, blockhostingvolume string) (
+
+	// if any of the cases above set err, log it and return
+	if err != nil {
+		logger.LogError("%v", err)
```
At least give some context to the person who might grep the code for an error line seen in the logs.

phlogistonjohn commented on executors/executor.go:
```diff
@@ -41,6 +41,8 @@ type Executor interface {
 	BlockVolumeCreate(host string, blockVolume *BlockVolumeRequest) (*BlockVolumeInfo, error)
 	BlockVolumeDestroy(host string, blockHostingVolumeName string, blockVolumeName string) error
+	BlockVolumeExpand(host string, blockHostingVolumeName string, blockVolumeName string, newSize int) error
+	InfoBlockVolume(host string, blockhostingvolume string, blockVolumeName string) (*BlockVolumeInfo, error)
```
BlockVolumeInfo

phlogistonjohn commented:
I just noticed that it is not in the TODO list you have, but we will need a error handling functional test at some point. Let me know if that something you need assistance with and/or want to do in a follow-up pr.

pkalever commented on apps/glusterfs/app_block_volume.go:
```diff
@@ -176,3 +176,39 @@ func (a *App) BlockVolumeDelete(w http.ResponseWriter, r *http.Request) {
+		logger.LogError("validation failed: " + err.Error())
+		return
+	}
+
```
ok

pkalever commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,312 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+			return ErrNoSpace
+		}
+
+		err = bve.bvol.expandBlockVolumeComponents(txdb, bve.newSize)
```
Perfectly makesense to me. Will take care.

pkalever commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,312 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+		if e := bve.op.Save(tx); e != nil {
+			return e
+		}
+		if e := bve.bvol.Save(tx); e != nil {
```
Got it.

pkalever commented:
@phlogistonjohn 

Split your latest comments into 3 different patches:
* V4: split updating the size of the bhv and bv
* V4: minimize the state carried across function calls
* V4: addressing comments from John   ----> minor code style comments

I hope I had addressed all of them, incase If I had missed anything, please let me know.


TODO for the next PR:
* Write Unit tests (including one for toGigaBytes() Function)
* error handling functional test


Thank you! For the best-detailed review.

phlogistonjohn commented on apps/glusterfs/block_volume_entry.go:
```diff
@@ -320,6 +320,44 @@ func (v *BlockVolumeEntry) destroyFromHost(
+	return blockVolumeInfo, nil
+}
+
+func (bv *BlockVolumeEntry) expandBlockVolumeComponents(db wdb.DB, newSize int) error {
```
The more I look at this function the more I think we should get rid of it. It is only called in one place, the name is misleading (IMO), and it mutates the db state of the parent BHV *only* not the block volume the method is connected to.
I suggest either moving the valuable logic into a private helper function in the operation or just remove it altogether and merge the logic where needed in the operation.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
 
+// BlockVolumeExpandOperation implements the operation functions used to
+// expand an existing blockvolume.
+type BlockVolExpandReclaim struct {
```
Let's lowercase this name to ensure future readers of the code that's it is private to the app logic.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+
+	// modification values
+	newSize int
+	Reclaim BlockVolExpandReclaim // gets set by Clean() call
```
Lowercase "reclaim" too.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+			return err
+		}
+		if bhv.Pending.Id != "" {
+			logger.LogError("Pending volume %v can not expand block volume %v", bhv.Info.Id, bve.bvolId)
```
The wording here is a bit awkward. How about: "Can not expand block volume %v when hosting volume %v is pending"

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+		if bve.newSize == bv.Info.Size {
+			logger.LogError("Requested new-size %v is same as current block volume size %v, nothing to be done.",
+				bve.newSize, bv.Info.Size)
+			return fmt.Errorf("Invalid Request: Requested new-size %v is same as current "+
```
Just return the result of logger.LogError [1] here, that text is (a) better/more clearly worded (b) the message the user can see on the CLI will be more easlily matched up with content in the logs.

1 - IIRC, LogError returns an `error` you can just use. If I'm misremembering create the error, log it, and then return it.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+		} else if bve.newSize < bv.Info.Size {
+			logger.LogError("Requested new-size %v is less than current block volume size %v, shrinking is not allowed.",
+				bve.newSize, bv.Info.Size)
+			return fmt.Errorf("Invalid Request: Requested new-size %v is less than current "+
```
Same as the above block.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+
+func (bve *BlockVolumeExpandOperation) Finalize() error {
+	return bve.db.Update(func(tx *bolt.Tx) error {
+		var bv *BlockVolumeEntry
```
This line is redundant.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+		if err != nil {
+			return err
+		}
+		actualSize = bvolInfo.Size
```
Unless you have a reason not to you could just assign to `bve.Reclaim.actualSize` here and skip the intermediate variable. But it's not a big deal.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,295 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+			if err != nil {
+				return err
+			}
+			ReclaimHsize := bve.newSize - bve.Reclaim.actualSize
```
Don't capitalize this. Also, what's the "H" mean?

phlogistonjohn commented on executors/cmdexec/block_volume.go:
```diff
@@ -192,3 +194,163 @@ func (c *CmdExecutor) ListBlockVolumes(host string, blockhostingvolume string) (
+	EXABYTE
+)
+
+func toGigaBytes(s string) (int, error) {
```
Since this function is critical to the proper functioning of this feature I must insist that we have some kind of unit test for this in this PR.

phlogistonjohn commented:
@raghavendra-talur PTAL

pkalever commented:
@phlogistonjohn updated with the patch fixing the latest review comments. I'm yet to add the Unit Test for toGigaBytes(), which is pending for this PR.

I will start working on the Unit Test for toGigaBytes() soon.

Thanks!

pkalever commented:
@phlogistonjohn @raghavendra-talur added unit test for toGigaBytes()

Thanks!

phlogistonjohn commented:
> @phlogistonjohn @raghavendra-talur added unit test for toGigaBytes()
> 
> Thanks!

Test looks good with one line uncovered, which is the default case for the switch statement. If you add another section to the test that creates a bogus suffix I think you will trigger that error case and get full coverage of the function. Thanks!

pkalever commented:
> 
> Test looks good with one line uncovered, which is the default case for the switch statement. If you add another section to the test that creates a bogus suffix I think you will trigger that error case and get full coverage of the function. Thanks!

Thanks @phlogistonjohn , updated now.

phlogistonjohn commented:
One additional request not related directly to the code:
In heketi we generally try to have commits be broken up by functional change rather than reflect the history of the code review itself. I'm not going to ask you to rewrite the history to make that style but I would like to squash the change-sets that only reflect evolution of the code during review. So for simplicity, I suggest just squashing everything here into once patch. Thanks.

pkalever commented:
@phlogistonjohn I understand. I will do the squashing part anyway once you are satisfied with the code.

When code changes are complex it makes reviewer job easy to add new review comment addressing as new patches. Of course it will be an additional responsibility of contributions to reclassify the changes into different parts. But IMHO it's worth it.

So once you say the changes look good, I will split the changes to different patches as needed.

pkalever commented:
@phlogistonjohn would you like me to squash them and resubmit?

Thanks!

phlogistonjohn commented:
Yes please.

pkalever commented:
@phlogistonjohn squashed them now. Travis CI crashing everytime on s390x/Go-1.1.3 attempting to run "# go get github.com/mattn/goveralls", doesn't look related to this PR anyways.

Thanks!

phlogistonjohn commented:
Yeah. I'm thinking about changing the travis config and just disabling s390.

pkalever commented:
> Thank you for the contribution. I look forward to seeing the follow-up PRs with the functional tests, etc!

Yes @phlogistonjohn, as agreed before, I will spend some time this week to work on the pending list and submit them in a fresh PR.

BTW the items in the pending TODO list are:
* Writing Unit tests
* Writing error handling functional tests

raghavendra-talur commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,288 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+			return err
+		}
+
+		requiredFeeSize := bve.newSize - bv.Info.Size
```
Small typo but it also gives a new meaning to the statement.

pkalever replied:
ok

raghavendra-talur commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,288 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+				return err
+			}
+		}
+		if err := bve.op.Save(tx); err != nil {
```
Should this be a delete of the operation instead of a save?

phlogistonjohn replied:
I agree. This does look wrong to me now that you point it out. Let's fix this now.
This is another reason why I don't love doing the tests in a seperate PR because we should have a check in the tests that ensures no pending ops after the clean + clean done functions are run. Still not going to insist on that but this is why we do tests and not just reviews. :-)

pkalever replied:
Good catch.

phlogistonjohn commented on apps/glusterfs/operations_block_volume.go:
```diff
@@ -376,6 +376,288 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+				return err
+			}
+		} else {
+			logger.Info("Actual Size match newly requested Size %v", bve.reclaim.actualSize)
```
Just noticed a minor grammar error here. Should be " Actual Size matches newly requested Size "

pkalever replied:
ok

pkalever commented:
@phlogistonjohn @raghavendra-talur updated the patch addressing the review comments. Once the reviewing is done, will squash it and resend.

Thanks!

phlogistonjohn commented:
Latest changes look OK to me. Please squash at your earliest convenience.

pkalever commented:
> Latest changes look OK to me. Please squash at your earliest convenience.

@phlogistonjohn done now. Thanks!

phlogistonjohn commented:
retest this please

phlogistonjohn commented:
retest this please
